File: aoc2023-rust/src/day2.rs
// https://adventofcode.com/2023/day/2

use std::fs;

struct Cube {
    red: u32,
    green: u32,
    blue: u32,
}

pub fn run() {
    let content = fs::read_to_string("../../advent-of-code-data/aoc2023/Day2.txt").unwrap();

    let mut sum_ids = 0;

    for line in content.lines() {
        let fields: Vec<_> = line.split(": ").collect();
        let gameid: u32 = fields[0].split(" ").collect::<Vec<_>>()[1].parse().unwrap();

        if is_valid_game(fields[1]) {
            sum_ids += gameid;
        }
    }

    println!("Day2: {}", sum_ids);
}

fn is_valid_game(inp: &str) -> bool {
    let mut is_valid = true;

    for cs in inp.split("; ") {
        let mut cube = Cube {
            red: 0,
            green: 0,
            blue: 0,
        };

        for c in cs.split(", ") {
            let mut it = c.split(" ");
            let n: u32 = it.next().unwrap().parse().unwrap();

            let color = it.next().unwrap();
            match color {
                "red" => cube.red += n,
                "green" => cube.green += n,
                "blue" => cube.blue += n,
                _ => panic!("Invalid color"),
            }
        }

        is_valid &= cube.red <= 12 && cube.green <= 13 && cube.blue <= 14;
    }

    is_valid
}

File: aoc2023-zig/src/day2.zig
// https://adventofcode.com/2023/day/2

const std = @import("std");
const Allocator = std.mem.Allocator;
const assert = std.debug.assert;

const Cube = struct {
    red: u32,
    green: u32,
    blue: u32,
};

pub fn run(allocator: Allocator) !void {
    var file = try std.fs.cwd().openFile("../../advent-of-code-data/aoc2023/Day2.txt", .{});
    defer file.close();

    const stat = try file.stat();

    const buffer = try file.readToEndAlloc(allocator, stat.size);
    defer allocator.free(buffer);

    var lines = std.mem.splitScalar(u8, buffer, '\n');

    var id_sum: u32 = 0;

    while (lines.next()) |line| {
        if (line.len == 0) {
            continue;
        }

        if (try isValidGame(line)) |id| {
            id_sum += id;
        } else {
            // ignore games that aren't valid
        }
    }

    std.debug.print("Day2: {d}\n", .{id_sum});
}

fn isValidGame(line: []const u8) !?u32 {
    var fields = std.mem.splitSequence(u8, line, ": ");
    const game_id = try std.fmt.parseInt(u32, fields.next().?[5..], 10);

    var cubes = std.mem.splitSequence(u8, fields.next().?, "; ");

    var is_valid = true;

    while (cubes.next()) |cube_| {
        var colors = std.mem.splitSequence(u8, cube_, ", ");

        var cube = Cube{ .red = 0, .green = 0, .blue = 0 };

        while (colors.next()) |color| {
            var it = std.mem.splitSequence(u8, color, " ");
            const n = try std.fmt.parseInt(u32, it.next().?, 10);
            const val = it.next().?;

            if (std.mem.eql(u8, val, "red")) {
                cube.red += n;
            } else if (std.mem.eql(u8, val, "green")) {
                cube.green += n;
            } else if (std.mem.eql(u8, val, "blue")) {
                cube.blue += n;
            } else {
                assert(false);
            }
        }

        is_valid = is_valid and isValidCube(cube);

        // std.debug.print("{d}     {any} {}\n", .{ game_id, cube, is_valid });
    }

    if (is_valid) {
        return game_id;
    } else {
        return null;
    }
}

fn isValidCube(cube: Cube) bool {
    return cube.red <= 12 and cube.green <= 13 and cube.blue <= 14;
}
